File: determinante.lua
-- Heap's Algorithm
local function permuta(A, lp)
	
	local n = #A
	
	local c = {}
	for i=1,n do
		c[i] = 1
	end
	
	index = 1
	lp[index] = {}
	for j=1,n do
		lp[index][j] = A[j]
	end
	
	
	local i = 2
	while i <= n do
		if c[i] < i then
			if i%2 == 1 then
				A[1], A[i] = A[i], A[1]
			else
				A[c[i]], A[i] = A[i], A[c[i]]
			end
			
			index = index + 1
			lp[index] = {}
			for j=1,n do
				lp[index][j] = A[j]
			end
			
			c[i] = c[i] + 1
			i = 2
		else
			c[i] = 1
			i = i + 1
		end
	end 
end

local function numberInversions(A)
	
	local s = 0
	for i=1,#A do
		for j=i,#A do
			if A[i] > A[j] then s = s +1 end
		end
	end
	return s%2
end


local function determinant(A)

	local det = 0
	
	local l = {}
	for i=1,#A do
		l[i] = i
	end
	
	local lp = {}
	
	permuta(l, lp)

	for _,__ in ipairs(lp) do
		local prod = 1
		local sign = numberInversions(__)
		for i,j in ipairs(__) do
			prod = prod * A[i][j]
		end
		det = det + ((-1)^sign)*prod
	end
	
	return det
end

matrix = {
	{1,12,4,-1},
	{4,5,-3,11},
	{-15,7,8,9},
	{5,12,45,67}
}

print(determinant(matrix))

File: determinante.py
# Heap's Algorithm
def permuta(A, lp):
	
	n = len(A)
	
	c = [0 for i in range(n)]
	aux = [A[i] for i in range(n)]
	
	lp.append(aux)

	i = 1
	while i < n:
		if c[i] < i:
			if i%2 == 0:
				A[0], A[i] = A[i], A[0]
			else:
				A[c[i]], A[i] = A[i], A[c[i]]
			aux = [A[i] for i in range(n)]		
			lp.append(aux)
			c[i] = c[i] + 1
			i = 1
		else:
			c[i] = 0
			i = i + 1

def numberInversions(A):
	
	s = 0
	for i in range(len(A)):
		for j in range(i,len(A)):
			if A[i] > A[j]: s = s +1

	return s%2
	
def determinant(A):

	det = 0

	l = [i for i in range(len(A))]
	lp = []
	
	permuta(l, lp)

	for _ in lp:
		prod = 1
		sign = numberInversions(_)
		for i,j in enumerate(_):
			prod = prod * A[i][j]
		det = det + ((-1)**sign)*prod
	
	return det


def main()
	matrix = [
		[1,2,4.5,-1],
		[4,5,-3,11],
		[-15,7,8,9],
		[5,12,45,67]
	]

	print(determinant(matrix))
	
	return



